File: C++/quick sort.1.cpp
#include <iostream>
#include <vector>
using namespace std;

int quickCount = 0; // Biến đếm số lần so sánh trong Quick Sort

int partition(vector<int>& arr, int low, int high) {
    int pivot = arr[high];
    int i = low - 1;

    for (int j = low; j < high; j++) {
        quickCount++;
        if (arr[j] <= pivot) {
            i++;
            swap(arr[i], arr[j]);
        }
    }
    swap(arr[i + 1], arr[high]);
    return i + 1;
}

void quickSort(vector<int>& arr, int low, int high) {
    if (low < high) {
        int p = partition(arr, low, high);
        quickSort(arr, low, p - 1);
        quickSort(arr, p + 1, high);
    }
}

int main() {
    vector<int> arr = {12, 11, 13, 5, 6, 7};
    quickSort(arr, 0, arr.size() - 1);
    cout << "Kết quả sau Quick Sort: ";
    for (int num : arr) {
        cout << num << " ";
    }
    cout << "\nSố lần so sánh trong Quick Sort: " << quickCount << endl;
    return 0;
}

File: python/quick sort.1.py
quick_count = 0  # Biến đếm số lần so sánh trong Quick Sort

def partition(arr, low, high):
    global quick_count
    pivot = arr[high]
    i = low - 1

    for j in range(low, high):
        quick_count += 1
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

def quick_sort(arr, low, high):
    if low < high:
        p = partition(arr, low, high)
        quick_sort(arr, low, p - 1)
        quick_sort(arr, p + 1, high)

if __name__ == "__main__":
    arr = [12, 11, 13, 5, 6, 7]
    quick_sort(arr, 0, len(arr) - 1)
    print("Kết quả sau Quick Sort:", arr)
    print("Số lần so sánh trong Quick Sort:", quick_count)